### ISSUE TITLE
AWS SecretManager - problem when unit testing

### ISSUE THREAD

**Issue by JCapriotti** [hynek/environ-config#24]
This is more of a question... possibly pointing to an issue relating to a documentation gap (which I could do a PR for).

When unit testing my project, I configure `environ-config` using a dictionary rather than `os.environ`, similar to this:

```python
environ.to_config(Configuration, {"SOME_VAR": "FOO"})
```

I got a little stuck after I added `SecretsManagerSecrets`, as it seems to always attempt to read from AWS Secrets Manager with boto3. At least, that's what I experienced with my particular setup.

I attempted to use `moto` and pass that client to `SecretsManagerSecrets`. However, I wasn't able to get that to work... it's possible my issues were due to me not using the library the way it is supposed to be used.... my rough setup is like this:

```python
from environ.secrets.awssm import SecretsManagerSecrets

sm: SecretsManagerSecrets = SecretsManagerSecrets()

@environ.config(prefix='')
class Configuration:
    client_id = environ.var()
    client_secret = sm.secret()
```

In the course of trying to get `moto` in there, I did some work to initialize `sm` in my code's "bootstrap" methods, and I also tried patching. 

But at the end of the day and after a bourbon, it seemed like it would be easier to simply set a default value:

```python
client_secret = sm.secret('unit_test_default')
```

So... I guess I have two questions
* Is there any high-level guidance for how `SecretsManagerSecrets` is supposed to be used? i.e. is my code reasonable :)
* If I'm on the right track, how is `SecretsManagerSecrets` intended to work with unit testing?

**Comment by offbyone** [comment]
Whoooof... 

(author of that particular module, here)

I have been racking my brain trying to understand why it's doing what you're seeing. I've written what _should_ be a working unit test for that behaviour -- it was missing in the original feature, so that's a miss! -- and I can clearly replicate the issue you're describing. I have no idea why though.

So, the short answer is, right now you are correct; you need a default. You _shouldn't_, but you do. I'm going to see if I can figure out why the client being created for the configuration isn't using the moto client.

**Comment by JCapriotti** [comment]
Thanks for looking so quickly!

I wasn't sure about this line in my code, it felt a little wrong:

```python
sm: SecretsManagerSecrets = SecretsManagerSecrets()
```

If I'm understanding the module, with that line it will use `boto3` by default. That's where I got into assuming I had to refactor the way it is initialized, patching, etc. Long story short, I wasn't sure the best way to get `moto` in there.

TBH I was hoping I could avoid `moto` altogether by calling something like this in my unit tests:

```python
environ.to_config(Configuration, {"CLIENT_ID": "foo", "CLIENT_SECRET": "bar"})
```

**Comment by offbyone** [comment]
`moto` patches the other clients, when you run your tests with `with mock_secretsmanager():` around them. But it's not working here, and I'm not sure why.

**Comment by offbyone** [comment]
Okay, I've figured out how to do it. If you are using `moto` you want roughly this:

```python
def test_a_config():
     sm = SecretsManagerSecrets()

     @environ.config
     class Cfg(object):
         pw = sm.secret()

    with mock_secretsmanager():
        # we need to make sure we're using the same region. It doesn't matter which -- moto _and_
        # boto will try figure it out from the environment -- but it has to be the same.
        sm.client.create_secret(Name="SecretName")
        sm.client.put_secret_value(
            SecretId="SecretName", SecretString="no-default"
        )
        conf = environ.to_config(Cfg, {"APP_PW": "SecretName"})
        assert conf.pw == "no-default"
```

The important thing is that you have to make sure that you put the secret you are testing in the same region as moto is going to use.

_However_, and this is important, right now the SecretsManagerSecret initializes its Boto client aggressively, and therefore cannot be mocked this way. Expect a PR in ... a few minutes that will address this.

**Comment by hynek** [comment]
fixed by #25
